### PR TITLE
feat: Add command line/environment flag for commonConfig

### PR DESCRIPTION
**Pull request by ejlee3**
If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [x] I have opened a PR for the related docs change (if not, why?)
[docs PR](https://github.com/edgexfoundry/edgex-docs/pull/990)

## Testing Instructions

**Non-secure mode testing **
1. set the environment variable for non-secure testing `export EDGEX_SECURITY_SECRET_STORE=false`
2. run the latest compose builder `make run no-secty` to start core services
3. stop core data
4. remove the core data configuration in consul
5. modify local go-mod in edgex-go to use bootstrap from [this branch](https://github.com/ejlee3/go-mod-bootstrap/tree/common-config-flag)
7. build core data `make data`
9. run core-data locally and ensure that the configuration loads `cd cmd/core-data` and `./core-data -cc ../core-common-config-bootstrapper/res/configuration.yaml`
10. Verify that the core data configuration and common config were loaded using curl `curl http://localhost:59880/api/v2/config | json_pp`

**secure mode testing **
Repeat the above steps with `EDGEX_SECRUITY_SECRETS_STORE=true`, run all services in secure mode, and run core data with `sudo`

### CODE REVIEW

File: bootstrap/config/config.go [hunk 1]
```diff
@@ -180,6 +181,13 @@ func (cp *Processor) Process(
 
 	// Now must load configuration from local file if any of these conditions are true
 	if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {
```
Comment (lenny-intel) [inline bootstrap/config/config.go:183]:
```suggestion
	if !useProvider {
```

File: bootstrap/config/config.go [hunk 2]
```diff
@@ -180,6 +181,13 @@ func (cp *Processor) Process(
...
+		commonConfigLocation := environment.GetCommonConfig(cp.lc, cp.flags.CommonConfig())
+		if commonConfigLocation != "" {
+			err := cp.loadCommonConfigFromUri(commonConfigLocation, serviceConfig, serviceType)
```
Comment (lenny-intel) [inline bootstrap/config/config.go:186]: Lets not use UIR term yet. When we do implement URI if will be a generic function not specific to configuration. My thought is this loadCommonConfigFromFile will can it to do the actually loading once we have it in place.
```suggestion
			err := cp.loadCommonConfigFromFile(commonConfigLocation, serviceConfig, serviceType)
```

File: bootstrap/config/config.go [hunk 3]
```diff
@@ -180,6 +181,13 @@ func (cp *Processor) Process(
...
+			if err != nil {
+				return fmt.Errorf("could not load oommon config: %s", err.Error())
```
Comment (lenny-intel) [inline bootstrap/config/config.go:188]: error already has context
```suggestion
				return err
```

File: bootstrap/config/config.go [hunk 4]
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
...
 
+// loadCommonConfigFromUri will pull up to two separate common configs from the config provider
+func (cp *Processor) loadCommonConfigFromUri(
+	configFile string,
+	serviceConfig interfaces.Configuration,
+	serviceType string) error {
+
+	var err error
+
+	//TODO: implement reading from URI
```
Comment (lenny-intel) [inline bootstrap/config/config.go:329]: The TODO can be removed

File: bootstrap/config/config.go [hunk 5]
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
...
+	commonConfig, err := cp.loadConfigYamlFromFile(configFile)
+	if err != nil {
+		return fmt.Errorf("could not load common configuration %s: %s", configFile, err.Error())
+	}
+	// separate out the necessary sections
+	allServicesConfig, ok := commonConfig["all-services"].(map[string]any)
+	if !ok {
+		return fmt.Errorf("could not find all-services section in config %s", configFile)
```
Comment (lenny-intel) [inline bootstrap/config/config.go:337]:
```suggestion
		return fmt.Errorf("could not find all-services section in common config %s", configFile)
```

File: bootstrap/config/config.go [hunk 6]
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
...
+	}
+	// use the service type to separate out the necessary sections
+	var serviceTypeConfig map[string]any
+	switch serviceType {
+	case config.ServiceTypeApp:
+		cp.lc.Infof("loading the common configuration for service type %s", serviceType)
+		serviceTypeConfig, ok = commonConfig[appServicesKey].(map[string]any)
+		if !ok {
+			return fmt.Errorf("could not find %s section in config %s", appServicesKey, configFile)
```
Comment (lenny-intel) [inline bootstrap/config/config.go:346]:
```suggestion
			return fmt.Errorf("could not find %s section in common config %s", appServicesKey, configFile)
```

File: bootstrap/config/config.go [hunk 7]
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
...
+		}
+	case config.ServiceTypeDevice:
+		cp.lc.Infof("loading the common configuration for service type %s", serviceType)
+		serviceTypeConfig, ok = commonConfig[deviceServicesKey].(map[string]any)
+		if !ok {
+			return fmt.Errorf("could not find %s section in config %s", deviceServicesKey, configFile)
```
Comment (lenny-intel) [inline bootstrap/config/config.go:352]:
```suggestion
			return fmt.Errorf("could not find %s section in common config %s", deviceServicesKey, configFile)
```

File: bootstrap/config/config.go [hunk 8]
```diff
@@ -310,6 +318,54 @@ func (cp *Processor) loadCommonConfig(
...
+	configFile string,
+	serviceConfig interfaces.Configuration,
+	serviceType string) error {
+
+	var err error
+
+	//TODO: implement reading from URI
+	commonConfig, err := cp.loadConfigYamlFromFile(configFile)
+	if err != nil {
+		return fmt.Errorf("could not load common configuration %s: %s", configFile, err.Error())
```
Comment (lenny-intel) [inline bootstrap/config/config.go:332]: error already has context
```suggestion
		return err
```

File: bootstrap/config/config_test.go [hunk 9]
```diff
@@ -246,6 +247,54 @@ func TestLoadCommonConfig(t *testing.T) {
...
+			// create the processor
+			proc := NewProcessor(f, env, timer, ctx, &wg, nil, dic)
+
+			// call load common config
+			err := proc.loadCommonConfigFromUri(tc.config, tc.serviceConfig, tc.serviceType)
+			// make assertions
+			require.NotNil(t, cancel)
+			if tc.expectedErr == "" {
+				assert.NoError(t, err)
+				assert.NotEmpty(t, tc.serviceType)
```
Comment (lenny-intel) [inline bootstrap/config/config_test.go:290]: Why is this assert needed? It is part of the tests code, not the code being tested.

File: bootstrap/environment/variables.go [hunk 10]
```diff
@@ -358,15 +359,27 @@ func GetProfileDir(lc logger.LoggingClient, profileDir string) string {
...
-		logEnvironmentOverride(lc, "-cf/--configFile", envConfigFile, envValue)
+		logEnvironmentOverride(lc, "-cf/--configFile", envKeyConfigFile, envValue)
 	}
 
 	return configFileName
 }
 
+// GetCommonConfig gets the common configuration value from the Variables value (if it exists)
+// or uses passed in value.
+func GetCommonConfig(lc logger.LoggingClient, commonConfig string) string {
```
Comment (lenny-intel) [inline bootstrap/environment/variables.go:373]:
```suggestion
func GetCommonConfigFileName(lc logger.LoggingClient, commonConfigFileName string) string {
```

File: bootstrap/environment/variables_test.go [hunk 11]
```diff
@@ -238,6 +238,35 @@ func TestGetConfigFileName(t *testing.T) {
 	}
 }
 
+func TestGetCommonConfig(t *testing.T) {
```
Comment (lenny-intel) [inline bootstrap/environment/variables_test.go:241]:
```suggestion
func TestGetCommonConfigFileName(t *testing.T) {
```

File: bootstrap/flags/flags.go [hunk 12]
```diff
@@ -133,6 +137,11 @@ func (d *Default) ConfigFileName() string {
 	return d.configFileName
 }
 
+// CommonConfig returns the uri for the common configuration
```
Comment (lenny-intel) [inline bootstrap/flags/flags.go:140]:
```suggestion
// CommonConfig returns the location for the common configuration
```

File: bootstrap/flags/flags.go [hunk 13]
```diff
@@ -145,6 +154,8 @@ func (d *Default) helpCallback() {
 			"Server Options:\n"+
 			"    -cp, --configProvider           Indicates to use Configuration Provider service at specified URL.\n"+
 			"                                    URL Format: {type}.{protocol}://{host}:{port} ex: consul.http://localhost:8500\n"+
+			"    -cc, --commonConfig             Takes the URI that specifies where the common configuration is pulled from when\n"+
```
Comment (lenny-intel) [inline bootstrap/flags/flags.go:157]:
```suggestion
			"    -cc, --commonConfig             Takes the location that specifies where the common configuration is loaded from when\n"+
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: bootstrap/flags/flags.go [hunk 14]
```diff
@@ -154,7 +154,7 @@ func (d *Default) helpCallback() {
 			"Server Options:\n"+
 			"    -cp, --configProvider           Indicates to use Configuration Provider service at specified URL.\n"+
 			"                                    URL Format: {type}.{protocol}://{host}:{port} ex: consul.http://localhost:8500\n"+
-			"    -cc, --commonConfig             Takes the URI that specifies where the common configuration is pulled from when\n"+
+			"    -cc, --commonConfig             Takes the location to specifies where the common configuration is loaded from when\n"+
```
Comment (lenny-intel) [inline bootstrap/flags/flags.go:157]: Takes the location to where the common configuration is loaded ....

File: bootstrap/config/config.go [hunk 15]
```diff
@@ -176,10 +177,15 @@ func (cp *Processor) Process(
...
-	if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {
+	} else {
+		// Now must load configuration from local file if any of these conditions are true
+		commonConfigLocation := environment.GetCommonConfigFileName(cp.lc, cp.flags.CommonConfig())
+		if commonConfigLocation != "" {
+			err := cp.loadCommonConfigFromFile(commonConfigLocation, serviceConfig, serviceType)
+			if err != nil {
+				return err
+			}
+		}
```
Comment (lenny-intel) [inline bootstrap/config/config.go:188]:
```suggestion
	} else {
		// Now must load configuration from local file if any of these conditions are true
		commonConfigLocation := environment.GetCommonConfigFileName(cp.lc, cp.flags.CommonConfig())
		if commonConfigLocation != "" {
			err := cp.loadCommonConfigFromFile(commonConfigLocation, serviceConfig, serviceType)
			if err != nil {
				return err
			}
		}
	}
	
       	// Now must load private configuration from local file if any of these conditions are true
       if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: bootstrap/config/config.go [hunk 16]
```diff
@@ -176,33 +177,42 @@ func (cp *Processor) Process(
...
 		}
 
-		// apply overrides - Now only done when loaded from file and values will get pushed into Configuration Provider (if used)
-		overrideCount, err := cp.envVars.OverrideTomlValues(tomlTree)
-		if err != nil {
-			return err
-		}
-		cp.lc.Infof("Configuration loaded from file with %d overrides applied", overrideCount)
+		// Now load the private config from a local file if any of these conditions are true
+		if !useProvider || !cp.providerHasConfig || cp.overwriteConfig {
```
Comment (lenny-intel) [inline bootstrap/config/config.go]: Move outside the else cause

**Review by lenny-intel — APPROVED** [review]
LGTM.

**Comment by MonicaisHer** [comment]
I updated the go-mod in edgex-go to use the bootstrap from this PR, but encountered some errors while building edgex-go. Here are the build steps and error details:

1. I added `replace github.com/edgexfoundry/go-mod-bootstrap/v3 v3.0.0-dev.54 => github.com/ejlee3/go-mod-bootstrap/v3 v3.0.0-20230320232408-c3ed78377b49` to the go.mod file and ran `go mod tidy`
2. I ran `snapcraft` to build the snapped binary
However, I received the following error:
```
internal/security/config/command/flags_common.go:37:9: cannot use &commonFlags (value of type *commonFlags) as flags.Common value in return statement: *commonFlags does not implement flags.Common (missing method CommonConfig)
make: *** [Makefile:134: cmd/secrets-config/secrets-config] Error 1
```
I'm wondering if there are any adjustments that need to be made to the edgex-go code as well?

File: bootstrap/config/config.go [hunk 17]
```diff
@@ -310,6 +326,53 @@ func (cp *Processor) loadCommonConfig(
...
+	serviceType string) error {
+
+	var err error
+
+	commonConfig, err := cp.loadConfigYamlFromFile(configFile)
+	if err != nil {
+		return err
+	}
+	// separate out the necessary sections
+	allServicesConfig, ok := commonConfig["all-services"].(map[string]any)
```
Comment (cloudxxx8) [inline bootstrap/config/config.go:342]: use `allServicesKey`

File: bootstrap/config/config.go [hunk 18]
```diff
@@ -527,10 +590,27 @@ func (cp *Processor) loadPrivateFromFile() (*toml.Tree, error) {
 		return nil, fmt.Errorf("could not convert to TOML Tree: %s", err.Error())
 	}
 
-	cp.lc.Info(fmt.Sprintf("Loaded private configuration from %s", filePath))
+	cp.lc.Infof(fmt.Sprintf("Loaded private configuration from %s", filePath))
 	return tomlTree, nil
 }
 
+// loadConfigYamlFromFile attempts to read the configuration yaml file
+func (cp *Processor) loadConfigYamlFromFile(yamlFile string) (map[string]interface{}, error) {
```
Comment (cloudxxx8) [inline bootstrap/config/config.go:598]: use `map[string]any` instead of `map[string]interface{}`

File: bootstrap/config/config.go [hunk 19]
```diff
@@ -527,10 +590,27 @@ func (cp *Processor) loadPrivateFromFile() (*toml.Tree, error) {
...
+	cp.lc.Infof("reading %s", yamlFile)
+	contents, err := os.ReadFile(yamlFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read common configuration file %s: %s", yamlFile, err.Error())
+	}
+
+	data := make(map[string]interface{})
```
Comment (cloudxxx8) [inline bootstrap/config/config.go:605]: use `map[string]any` instead of `map[string]interface{}`

**Comment by lenny-intel** [comment]
> I updated the go-mod in edgex-go to use the bootstrap from this PR, but encountered some errors while building edgex-go. Here are the build steps and error details:
> 
> 1. added `replace github.com/edgexfoundry/go-mod-bootstrap/v3 v3.0.0-dev.54 => github.com/ejlee3/go-mod-bootstrap/v3 v3.0.0-20230320232408-c3ed78377b49` to the go.mod file and ran `go mod tidy`
> 2. ran `snapcraft` to build the snapped binary
> 
> However, I received the following error:
> 
> ```
> internal/security/config/command/flags_common.go:37:9: cannot use &commonFlags (value of type *commonFlags) as flags.Common value in return statement: *commonFlags does not implement flags.Common (missing method CommonConfig)
> make: *** [Makefile:134: cmd/secrets-config/secrets-config] Error 1
> ```
> 
> I'm wondering if there are any adjustments that need to be made to the edgex-go code as well?

@MonicaisHer Yes, edgex-go will need tweaks for those services that re-implement the flags.Common interface. They need to implement the CommonConfig API.

**Review by lenny-intel — APPROVED** [review]
LGTM

**Comment by ejlee3** [comment]
> > I updated the go-mod in edgex-go to use the bootstrap from this PR, but encountered some errors while building edgex-go. Here are the build steps and error details:
> > 
> > 1. added `replace github.com/edgexfoundry/go-mod-bootstrap/v3 v3.0.0-dev.54 => github.com/ejlee3/go-mod-bootstrap/v3 v3.0.0-20230320232408-c3ed78377b49` to the go.mod file and ran `go mod tidy`
> > 2. ran `snapcraft` to build the snapped binary
> > 
> > However, I received the following error:
> > ```
> > internal/security/config/command/flags_common.go:37:9: cannot use &commonFlags (value of type *commonFlags) as flags.Common value in return statement: *commonFlags does not implement flags.Common (missing method CommonConfig)
> > make: *** [Makefile:134: cmd/secrets-config/secrets-config] Error 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm wondering if there are any adjustments that need to be made to the edgex-go code as well?
> 
> @MonicaisHer Yes, edgex-go will need tweaks for those services that re-implement the flags.Common interface. They need to implement the CommonConfig API.

Draft PR for edgex-go changes is here: https://github.com/edgexfoundry/edgex-go/pull/4464

**Comment by lenny-intel** [comment]
@MonicaisHer , please let us know once you have verified in edgex-go snap so we can merge this PR. THX!

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Review by MonicaisHer — APPROVED** [review]
Thank you! I tested this [PR](https://github.com/edgexfoundry/edgex-go/pull/4464) with snap to use a common config option, and it worked as expected. I was able to choose not to use the config provider and instead pull the common config from a file. The environment overrides were then applied after that. Here are the steps I used:
```
# Install snap built from this PR's changes
$ snap install --dangerous ./edgexfoundry_3.0.0-dev.108_amd64.snap

# Set not to use config provider
sudo snap set edgexfoundry apps.core-data.config.edgex-config-provider='none'

# Pull common config
sudo snap set edgexfoundry apps.core-data.config.edgex-common-config=./config/core-common-config-bootstrapper/res/configuration.yaml

# Set test environment variables
sudo snap set edgexfoundry apps.core-data.config.service-startupmsg='manual-snap-testing'

snap restart edgexfoundry.core-data
```
To verify that the common config was pulled and the environment variables were applied:
```
$ journalctl -f | grep core-data
app=core-data source=variables.go:402 msg="Variables override of 'Configuration Provider Information' by environment variable: EDGEX_CONFIG_PROVIDER=none"
app=core-data source=variables.go:402 msg="Variables override of '-cc/--commonConfig' by environment variable: EDGEX_COMMON_CONFIG=./config/core-common-config-bootstrapper/res/configuration.yaml
app=core-data source=config.go:599 msg="reading ./config/core-common-config-bootstrapper/res/configuration.yaml"
app=core-data source=config.go:194 msg="Common configuration loaded from file with 1 overrides applied"
app=core-data source=message.go:51 msg="Starting core-data 3.0.0-dev.108 "
app=core-data source=message.go:55 msg="manual-snap-testing"
```

**Pull request merged**